docs/DesignDocs/AttestationAPIsWithFlexibleClaims.md

```diff
@@ -0,0 +1,571 @@
+  ```c
+  /**
+   * oe_verify_attestation_certificate_with_evidence_v3
+   *
+   * This function is similar to
+   * oe_verify_attestation_certificate_with_evidence_v2 that has the same
+   * prototype but with the ability to verify a ceritifacte generated by
+   * oe_get_attestation_certificate_with_evidence_v3.
+   *
+   * @experimental
+   *
+   * @param[in] certificate_in_der A pointer to buffer holding certificate
+   * contents in DER format.
+   * @param[in] certificate_in_der_size The size of the buffer above.
+   * @param[in] endorsements_buffer A pointer to buffer holding endorsements.
+   * @param[in] endorsements_buffer_size The size of the buffer above.
+   * @param[in] policies A pointer to buffer holding policies.
+   * @param[in] policies_size The size of the buffer above.
+   * @param[out] claims A double-pointer to buffer holding claims.
+   * @param[out] claims_length The size of the buffer above.
+   * @retval OE_OK Successful validation.
+   * @retval OE_VERIFY_FAILED Quote failure.
+   * @retval OE_INVALID_PARAMETER One or more invalid parameters.
+   * @retval OE_FAILURE General failure.
+   * @retval Other appropriate error code.
+   */
+  oe_result_t oe_verify_attestation_certificate_with_evidence_v3(
+      const uint8_t* certificate_in_der,
+      size_t certificate_in_der_size,
+      const uint8_t* endorsements_buffer,
+      size_t endorsements_buffer_size,
+      const oe_policy_t* policies,
+      size_t policies_size,
+      oe_claim_t** claims,
+      size_t* claims_length);
+  ```
+  The API does not introduce changes to the function prototype. The purpose of
+  the API is to pair with `oe_get_attestation_certificate_with_evidence_v3`;
+  i.e., only the API is able to verify the certificated generated via
+  `oe_get_attestation_certificate_with_evidence_v3` with `runtime_custom_claims_buffer`
+  or `inittime_custom_claims_buffer` supplied.
+
+User Experience
+--------
+
+The following example demonstrates the use of proposed APIs.
+
+```c
+static const uint8_t inittime_custom_claims[] =
+  { 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+    0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f };
+
+static const uint8_t runtime_custom_claims[] =
+ { 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x66, 0x77,
+   0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff };
+
+...
+
+// Attester
+uint8_t* output_certificate = nullptr;
+size_t output_certificate_size = 0;
+
+...
+
+result = oe_get_attestation_certificate_with_evidence_v3(
+     &format,
+     (const unsigned char*)"CN=Open Enclave SDK,O=OESDK TLS,C=US",
+     private_key,
+     private_key_size,
+     public_key,
+     public_key_size,
+     (uint8_t*)runtime_custom_claims,
+     sizeof(runtime_custom_claims),
+     (uint8_t*)inittime_custom_claims,
+     sizeof(inittime_custom_claims),
+     NULL,
+     0,
+     &output_certificate,
+     &output_certificate_size);
+OE_CHECK(result);
+
+...
+
+Send output_certificate to verifier
```

Comment:
I'm having a hard time understanding the use case here.  I'd expect the certificate to go to the relying party, and only the evidence to go to the verifier.

Reply:
This is a simplified end-to-end example (i.e., the model used in the attested-tls sample)

Reply:
The attested tls sample has a tls session between an attester and a relying party

Reply:
I cannot find that there is a relying party involves in the attested tls sample. Can you point it out?